### PR TITLE
Minor improvements when iterating matching kNN docs

### PR DESCRIPTION
**Pull request by benwtrent**
The two minor performance improvements are around `Weight#count` and `Weight#scorer`.
`segmentStarts` is a monotonically increasing start for each scored document indexed by leaf-segment ordinal. Consequently, if the upper and lower segments are equivalent, that means no docs match for this segment.

Count is similarly calculated by the difference between upper and lower segmentStarts according to the segment ordinal.

### CODE REVIEW

**Review by javanna — COMMENTED** [review]
Would it be possible to add a test to verify these changes?

File: server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQuery.java [hunk 1]
```diff
@@ -72,6 +72,12 @@ public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float bo
             throw new IllegalStateException("This KnnScoreDocQuery was created by a different reader");
         }
         return new Weight(this) {
+
+            @Override
+            public int count(LeafReaderContext context) {
+                return segmentStarts[context.ord + 1] - segmentStarts[context.ord];
+            }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQuery.java:79]: does the test below still succeed if you remove this count implementation? This is maybe tricky to test, I wonder if we are testing that we override count.

**Review by javanna — APPROVED** [review]
Left a question on testing, not a biggie though.

File: server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQuery.java (hunk 1 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQuery.java:79]: IMO, we simply need to make sure that `count` is the same as the actual iterator count. Which is what the test is doing. Of course, the test passed before this.

**Pull request merged**